### PR TITLE
Implement paginated extrinsics

### LINKED ISSUE #926: [Private Graph] Create extrinsic single page generic data
Create an extrinsic to Create_or_update a paginated storage.

### Inputs:
- msa_id: MessageSourceId
- schema_id: SchemaId
- page: StoragePage
```

struct StoragePage {
  number: PageNumber,
  data: vec<u8>,
  prev_hash: Option<[u8; 32]> // or T::Hash
}
```

The data should be store under a child tree of the msa and map between `(schemaId,pageNumber`)  to `vec<u8>`

### Acceptance Criteria:
- a user should be able to create a new or update an existing page for a Paginated Storage
- unit tests and integration tests
- benchmarks

More details in [here](https://github.com/LibertyDSNP/frequency/pull/900/files)

Noes:
ping @aramikm for child tree branch

### PR DESCRIPTION
**Pull request by JoeCap08055**
# Goal
The goal of this PR is to implement extrinsics for upserting/removing pages in Paginated Stateful Storage.

Closes #926

# Checklist
- [ ] Chain spec updated
- [ ] Custom RPC OR Runtime API added/changed? Updated js/api-augment.
- [ ] Design doc(s) updated
- [x] Tests added
- [x] Benchmarks added
- [x] Weights updated

### CODE REVIEW

File: common/primitives/src/stateful_storage.rs
Comment (aramikm) [inline common/primitives/src/stateful_storage.rs:2]: It's less likely that we might have pages more than 256 but I think to future proof this for groups let's change it to `u16`

File: pallets/stateful-storage/Cargo.toml
Comment (aramikm) [inline pallets/stateful-storage/Cargo.toml:40]: We should use the same 1 item per line styling, It's easier to see the diffs

File: pallets/stateful-storage/src/lib.rs
Comment (aramikm) [inline pallets/stateful-storage/src/lib.rs:304]: I think you need to wrap your payload inside `Page<T::MaxPaginatedPageSizeBytes)::from(payload.into())` before storing

File: pallets/stateful-storage/src/types.rs
Comment (aramikm) [inline pallets/stateful-storage/src/types.rs:75]: should be `debug`?

File: common/primitives/src/stateful_storage.rs
Comment (JoeCap08055) [reply common/primitives/src/stateful_storage.rs:2]: Done.

File: pallets/stateful-storage/src/types.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/types.rs:75]: Meant to remove that.

File: pallets/stateful-storage/Cargo.toml
Comment (JoeCap08055) [reply pallets/stateful-storage/Cargo.toml:40]: Weird; must have been some auto-formatter hook. Fixed.

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:281]: let provider_key = ensure_signed(origin)?;
I would move this check to the top since it checks that this transaction's origin is signed.

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:285]: Code golf:
	```
	let schema = T::SchemaProvider::get_schema_by_id(schema_id).amp_err (|_|  Error::<T>::InvalidSchemaId)?;
	```

Doing this will allow you to get rid of unwrap() below and do:
```
ensure!(
				schema.payload_location == PayloadLocation::Paginated,
				Error::<T>::SchemaPayloadLocationMismatch
			);
```

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:329]: Seems like this block of code can be refactored into a helper method.  Both extrinsic seem to be doing these checks. By moving into a helper method it allows for it to be unit tested.


```
fn ensure_valid_schema_something() {
		let schema = T::SchemaProvider::get_schema_by_id(schema_id);
			ensure!(schema.is_some(), Error::<T>::InvalidSchemaId);
			ensure!(
				schema.unwrap().payload_location == PayloadLocation::Paginated,
				Error::<T>::SchemaPayloadLocationMismatch
			);

			let current_block = frame_system::Pallet::<T>::block_number();
			T::SchemaGrantValidator::ensure_valid_schema_grant(
				ProviderId(provider_msa_id),
				DelegatorId(state_owner_msa_id),
				schema_id,
				current_block,
			)
			.map_err(|_| Error::<T>::UnAuthorizedDelegate)?;
			....
			}
```

**Comment by enddynayn** [comment]
Looking great!

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:305]: This extrinsic name contains `upsert` in the name. So, perhaps, we should consider calling this event PaginatedPageUpsert?

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:349]: To follow the CRUD naming conventions, perhaps we call this event Event::PaginatedPageDeleted?

File: pallets/stateful-storage/src/benchmarking.rs
Comment (saraswatpuneet) [inline pallets/stateful-storage/src/benchmarking.rs:42]: we can delete this and also remove the create_schema helper function looks like its not used?

**Review by saraswatpuneet — COMMENTED** [review]
lgtm [including comments from reviewers]

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:285]: But that method doesn't return a `Result`, it returns an `Option`. But you got me thinking--a little checking, and I can use `.ok_or()` instead of `.map_err()`. Thanks.

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:329]: Nice idea; I like it. Actually I think the same treatment can be applied to most of the checks in the extrinsics for this pallet.

File: pallets/stateful-storage/src/lib.rs
Comment (saraswatpuneet) [reply pallets/stateful-storage/src/lib.rs:349]: We can either create a chore for this if we want to handle this once all pieces are in ? But since there are no downstream consumers as of yet, I am ok with either or

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:281]: I've refactored this into a `check_origin` method that also gets the `MessageSourceId`. But I'm not calling it first thing because that requires a DB read; I figured better to check the payload first that doesn't require any DB reads.

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:340]: I see what you did here. Though it seems that ensure_valid_msa_key contributes to checking the origin, they are two different things.

When one checks the origin, it simply checks that the origin enum is a `Signed` origin. It does not do any DB reads. So it would be best not to wrap these together or change the name of this function. I do not have any recommendations for naming at the moment.

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs]: Take a look at the latest commit; refactored function `check_origin` does both signature verification AND looks up MSA (incurs DB read). If we really want signature verification to come first ALWAYS, I can rework it a little.

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/lib.rs:359]: I wonder if removing this call is ok:
```
let schema =
			T::SchemaProvider::get_schema_by_id(schema_id).ok_or(Error::<T>::InvalidSchemaId)?;
		ensure!(
			schema.payload_location == payload_location,
			Error::<T>::SchemaPayloadLocationMismatch
		);
```

It might be that we do not need it if you can grant that schema implies that the grant exists?

```
T::SchemaGrantValidator::ensure_valid_schema_grant(
			ProviderId(provider_msa_id),
			DelegatorId(state_owner_msa_id),
			schema_id,
			current_block,
		)
		```

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:340]: Gotcha.

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:359]: Problem is we need to verify the payload location matches. That information only exists on the schema itself, we can't get or infer it from the grant.

File: pallets/stateful-storage/src/benchmarking.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/benchmarking.rs:49]: Nice. I am guessing your updating the weights on Jenkins and weighting for them to complete to wire this up?

File: pallets/stateful-storage/src/lib.rs
Comment (enddynayn) [reply pallets/stateful-storage/src/lib.rs:359]: ahhh, yeah your right.

File: pallets/stateful-storage/src/benchmarking.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/benchmarking.rs:49]: I'm running benchmarks locally to check, but I think we're holding off on Jenkins and weighting until just before we merge the whole `private-graph-main` branch to `main`

File: pallets/stateful-storage/src/benchmarking.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/benchmarking.rs]: Hmmm... okay, I guess `create_schema` is necessary when actually running benchmarks (since it just failed on me locally). Updating...

File: pallets/stateful-storage/src/benchmarking.rs
Comment (enddynayn) [reply pallets/stateful-storage/src/benchmarking.rs:49]: It might not be a bad idea to generate this and check them since you already created the benchmark.

File: pallets/stateful-storage/src/mock.rs
Comment (enddynayn) [inline pallets/stateful-storage/src/mock.rs:175]: I would avoid mocking these out.

Comment (saraswatpuneet) [reply pallets/stateful-storage/src/mock.rs:175]: Agree with enddy, ideally  that would give us a complete roundtrip similar to main runtime

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs]: Done.

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs]: OK, refactored once more. Should be good now.

File: pallets/stateful-storage/src/benchmarking.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/benchmarking.rs]: Fixed.

File: pallets/stateful-storage/src/lib.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/lib.rs:340]: Refactored.

File: pallets/stateful-storage/src/mock.rs
Comment (JoeCap08055) [reply pallets/stateful-storage/src/mock.rs:175]: Hmmm... isn't the point of unit testing to only execute the unit under test and mock all external dependencies? (btw, this is how `pallet_messages` is doing it; I just copied the approach). Though personally I think we could benefit from test-specific implementations of `SchemaBenchmarkHelper` and `SchemaProvider` that kept track of created schemas using a BTreeMap or HashMap, so then we wouldn't have to do this hokey stuff. But I don't like the idea of taking a dependency on the Schemas pallet for unit testing.

Comment (saraswatpuneet) [reply pallets/stateful-storage/src/mock.rs:175]: Right, then we have to agree all such tests should live in integration tests ? I think we can definitely investigate this as a spike, since many other pallets would benefit from it like messages pallet which depends on schema as well as msa etc.

**Review by aramikm — APPROVED** [review]
Approved! I'll take care of outstanding issues and questions in my PR https://github.com/LibertyDSNP/frequency/pull/1041 to keep this moving.

**Pull request merged**
